src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs

```diff
@@ -44,12 +45,42 @@ impl StreamDynamicFilter {
             false, // TODO(rc): decide EOWC property
             watermark_columns,
         );
-        Self { base, core }
+        let cleaned_by_watermark = Self::cleaned_by_watermark(&core);
+        Self {
+            base,
+            core,
+            cleaned_by_watermark,
+        }
     }
 
     pub fn left_index(&self) -> usize {
         self.core.left_index()
     }
+
+    /// 1. Check the comparator.
+    /// 2. RHS input has to be temporal filter.
+    pub fn cleaned_by_watermark(core: &DynamicFilter<PlanRef>) -> bool {
+        let expr = core.predicate();
+        if let Some(ExprImpl::FunctionCall(function_call)) = expr.as_expr_unless_true() {
+            match function_call.func_type() {
+                ExprType::GreaterThan | ExprType::GreaterThanOrEqual => {
+                    let rhs_input = core.right();
+                    if let Some(stream_exchange) = rhs_input.as_stream_exchange() {
```

Comment:
Can just check the `rhs_input.watermark()` here

Reply:
I couldn't find this function (`watermark()`). Could you please elaborate?

Reply:
Do you mean `!rhs_input.watermark_columns().is_empty()`

Reply:
Makes sense to me. We could have other source of watermark.